lib/archive_importer.rb

```diff
@@ -0,0 +1,105 @@
+# frozen_string_literal: true
+
+class ArchiveImporter
+  include ArchiveHelper
+  include Diaspora::Logging
+
+  attr_accessor :user
+
+  def initialize(archive_hash)
+    @archive_hash = archive_hash
+  end
+
+  def import
+    ActiveRecord::Base.transaction do
```

Comment:
So far I fixed all the notes except of this one. This one requires some more thinking and discussion. At this point I should do some performance tests to see how it goes.

This also rises some questions. If a user tries to import using UI and sees that something goes wrong, then it is possible that the user will go to another pod and try there again. In this case we will get in a weird state where two pods may have different opinions about data belonging. Some of the data will be already imported for the new user on the first pod and the second pod will have different author for the same imported data. If we allow retrying then how do we handle this case?

Actually, during the account import phase no ownership transition happens. All of the data which is imported from the archive is new to the pod. And the new data is imported with the ownership set to the newly created user. The data which is already known to the pod is transfered to the new user during the migration phase which starts only after successful data import.

So, basically everything before the account migration phase (user creation and import phase) is revertible. No other pod gets any information about the migration and all local data changes are made around the new user DB entry. In this case we can reverse everything without DB transactions, by just destroying the new user object (and therefore all of the imported data). We have to correctly handle all of the database records relations for that.

So if some unexpected issue happens which leads to a failure we can just destroy all of the imported data and thus get back to the original point. This way it will be retryable and we avoid the weird state of "only some of the data imported".

Another thought came to my mind. Since import takes some time some impatient users may want to try to discover the account before the import is finished. If the user is discovered we can't destroy it because we'll make the username unavailable forever. So maybe we should destroy only imported data related to the new user, but not the user itself, and when retry is requested just reuse the same user which was created during the previous import.

What do you think?

Reply:
> Some of the data will be already imported for the new user on the first pod and the second pod will have different author for the same imported data.

That's very true, but nobody prevents you to import your archive to two pods in parallel which would end in the same scenario, or import the same archive later on a pod which didn't receive the migration. But I don't think we can prevent this scenario completely.

> So if some unexpected issue happens which leads to a failure we can just destroy all of the imported data and thus get back to the original point.

That's true too. But what happens when the delete of everything then fails too, then we have a half imported and half-half-deleted user ;) I think the risk for that is pretty small, but still there. I think when we want to undo it, we should probably do it with a transaction (but with benchmark-tests, to make sure how much ram it uses).

> What do you think?

The more I think about it, the more I think we should add support for account merging. That's something useful anyway (so I want to support that at some point), but I think it's already useful now. It would save most of our problems.

If the import fails, you can just import the same archive again to the same user, we don't need to undo anything. We maybe even could do the migration of existing data to the new user if the import fails. And then the user can retry to import missing data again later.

That way we also don't need to prevent the imported user to be discovered or interacted with. It can be a full working user even while the import/migration is still running, the user could even already start writing posts/comments.

The only problem it doesn't solve is, when somebody imports the same archive to two pods (either in parallel or one after the other). But I think we should just display a warning, that you shouldn't import it to another pod after the validation is done and the user starts the real import. I don't see any technical solution to prevent this otherwise.

Also: The problem with "merging" is, the current migration from 0.7.0.0 doesn't support it, but I think we should add support for it in a minor to 0.7.x (in a separate PR), because it makes a lot of the other things easier and solves possible problems without it. I already came to that solution a few times in the past, but the "it isn't in 0.7.x" blocked me, but as said, the more I think about it, I think it's the way to go.

But I think with adding it to a minor in 0.7.x it would be OK. Most posts update soon and even when a pod doesn't update a normal migration would still work, if the new user isn't discovered before (which is probably the case for most migrations). So I think it would still be OK to do it with a shorter period between adding support to 0.7.x and finishing the migration.

Back to you: What do you think? ;)

Reply:
> I think we should add support for account merging.

I also would love to see account merging being possible (if by which you mean the migration of one account into another existing account). But this could be handled in a future iteration if it will add complexity. The priority is, I think, to have a properly functioning migration feature as soon as possible. It can then be refined.

Reply:
> But this could be handled in a future iteration if it will add complexity.

@goobertron The thing is, it would add other complexity that we wouldn't need to do when we add merging now. And we want merging sometimes in the future anyway, so why not now? Also adding the extra complexity needed for merging would make the normal migration much more stable and failsafe.

Reply:
Thanks for your response. If it can be added now, I am MORE than happy!

Reply:
So I crafted [this](https://framadrop.org/r/NmxGBdvrlS#AiSlP7pROcZLNCqJu163fbnVWkUNrKfAyGWl+Q5KPr0=) archive, which contains 10000 posts. 

I ran the archive import and logged the size of residential memory set in the end of the process.

With `transaction` the size of residential memory set is `807596` and grows with the number of posts. Without `transaction` the size of residential memory set is `321308` and there doesn't seem to grow proportionally.

So while `807596` is not that bad result, we don't necessarily have to use the `transaction` because we support account merging now and because there is a theoretical option to revert the import by just deleting the imported content.